### PR TITLE
Add basic WebAssembly JS API example

### PR DESCRIPTION
**Pull request by yurydelendik**
Just some basic example to start working with WebAssembly in the SpiderMonkey. Ideally we want better JS_PUBLIC_API calls.

### CODE REVIEW

File: examples/wasm.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  if (!JS_GetProperty(cx, wasmObj, "Module", &wasmModule)) return false;
+  if (!JS_GetProperty(cx, wasmObj, "Instance", &wasmInstance)) return false;
+
+
+  // Construct Wasm module from bytes.
+  JS::RootedObject module_(cx);
+  {
+    JS::RootedValueArray<1> args(cx);
+    args[0].setObject(
+      *JS::NewArrayBufferWithUserOwnedContents(cx, hi_wasm_len, hi_wasm));
```
Comment (ptomato) [inline examples/wasm.cpp]: The return value of `JS::NewArrayBufferWithUserOwnedContents()` should be checked for null and bail out if so.

File: examples/wasm.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    if (!Construct(cx, wasmModule, args, &module_)) return false;
+  }
+
+  // Construct Wasm module instance with required imports.
+  JS::RootedObject instance_(cx);
+  {
+    // Build "env" imports object.
+    JS::RootedObject envImportObj(cx, JS_NewPlainObject(cx));
```
Comment (ptomato) [inline examples/wasm.cpp:79]: Same here, the return value of `JS_NewPlainObject()` should be checked for null and bail out if so.

File: examples/wasm.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    if (!JS_DefineFunction(cx, envImportObj, "bar", BarFunc, 1, 0)) return false;
+    JS::RootedValue envImport(cx, JS::ObjectValue(*envImportObj));
+    // Build imports bag.
+    JS::RootedObject imports(cx, JS_NewPlainObject(cx));
```
Comment (ptomato) [inline examples/wasm.cpp:84]: Same here.

**Review by ptomato — COMMENTED** [review]
This seems like a really useful example to have! I'm not so familiar with WASM myself so maybe someone else can comment on that part of it?

**Review by lars-t-hansen — APPROVED** [review]
Looks plausible.

**Pull request merged**